---
title: "R Genomics: RStudio and R Basics"
output:
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: inline
  markdown: 
    wrap: 72
---

# Introducing R and RStudio IDE

## Getting to work with R: navigating directories

Find out where you are using `getwd()`

```{r where_am_i}
# Find out where you are using getwd()
getwd()
```

[Organize your data and
code](https://kbroman.org/steps2rr/pages/organize.html)

Use function `dir.create()` to create the directories

-   R
-   data
-   rawdata

```{r create_dir}
# Use function dir.create() to create the directories
```

## Download raw data file

Use function `download.file()` to download the file

<https://ndownloader.figshare.com/files/14632895> to
rawdata/combined_tidy_vcf.csv

```{r download_data}
# Use function download.file() to download the file https://ndownloader.figshare.com/files/14632895 to rawdata/combined_tidy_vcf.csv
download.file(url = "https://ndownloader.figshare.com/files/14632895",destfile = "rawdata/combined_tidy_vcf.csv")
```

## Using functions in R, without needing to master them

There are a few ways to get help with an R function

1\. `?args()`

2.  use`args()` function around a function`args(round())`

```{r getting_help}
# get help with args function
#?
#args()
?getwd()
args(round)
```

## RStudio contextual help

RStudio also provides contextual help. Start typing function `round()`
and use tab

```{r rstudio_help}
# get contextual help for round()
# round(3.14)
# round(3.14,1)
```

# R Basics

## Objectives

-   Be able to create the most common R objects including vectors

-   Understand that vectors have modes, which correspond to the type of
    data they contain

-   Be able to use arithmetic operators on R objects

-   Be able to retrieve (subset), name, or replace, values from a vector

-   Be able to use logical operators in a sub-setting operation

-   Understand that lists can hold data of more than one mode and can be
    indexed

## **What these lessons will not teach you**

**Here are some R skills we will *not* cover in these lessons**

-   How to create and work with R matrices and R lists

-   How to create and work with loops and conditional statements, and
    the "apply" family of functions (which are super useful, read more
    [here](https://www.r-bloggers.com/r-tutorial-on-the-apply-family-of-functions/))

-   How to do basic string manipulations (e.g. finding patterns in text
    using grep, replacing text)

-   How to plot using the default R graphic tools (we *will* cover plot
    creation, but will do so using the popular plotting package
    `ggplot2`)

-   How to use advanced R statistical functions

## Creating objects in R

What might be called a `variable` in many languages, such as python, is
called an **object** in R.

**To create an object you need:**

-   a name (e.g. `a`)

-   a value (e.g. `1`)

-   the assignment operator (`<-`) you can also use an `=`

This line creates the object 'a' and assigns it the value '1'

`<-` `->` `=`

```{r}
# creates the object 'a' and assigns it the numeric value '1', <-, ->, =
a <-1
1 -> a
a=1
a
# create an object gene_name with the string value PTEN for gene name
# strings have quotes, single or double
gene_name="PTEN"
```

## **Exercise: Create some objects in R**

```{r}
# Create an object that has the value of number of pairs of human chromosomes
# If not already create an object gene_name with the value of a gene symbol
chr_num=23
```

## **Naming objects in R**

Here are some important details about naming objects in R.

-   **Avoid spaces and special characters**: Object names cannot contain
    spaces or the minus sign (`-`). You can use '\_' to make names more
    readable. You should avoid using special characters in your object
    name (e.g. ! \@ \# . , etc.). Also, object names cannot begin with a
    number.

-   **Use short, easy-to-understand names**: You should avoid naming
    your objects using single letters (e.g. 'n', 'p', etc.). This is
    mostly to encourage you to use names that would make sense to anyone
    reading your code (a colleague, or even yourself a year from now).
    Also, avoiding excessively long names will make your code more
    readable.

-   **Avoid commonly used names**: There are several names that may
    already have a definition in the R language (e.g. 'mean', 'min',
    'max'). One clue that a name already has meaning is that if you
    start typing a name in RStudio and it gets a colored highlight or
    RStudio gives you a suggested autocompletion you have chosen a name
    that has a reserved meaning.

-   **Use the recommended assignment operator**: In R, we use '\<- '

```{r}
# example of bad names
1val=1
val?q=1
first val=1

```

## **Reassigning object names or deleting objects**

Once an object has a value, you can change that value by overwriting it.

[**R will not give you a warning or error**]{.ul} if you overwriting an
object, which may or may not be a good thing depending on how you look
at it.

```{r}
# print contents of object gene _name
# gene_name has the value 'pten' or whatever value you used in the challenge.
# We will now assign the new value 'TP53'
gene_name
gene_name="TP53"
gene_name
```

You can also remove an object from R's memory entirely. The `rm()`
function will delete the object.

```{r}
# delete the object 'gene_name'
rm(gene_name)
```

If you run a line of code that has only an object name, R will normally
display the contents of that object.

In this case, we are told the object no longer exists.

```{r}
# print content of gene_name to screen
gene_name
```

## Understanding object data types (modes)

In R, every object has two properties:

-   `Length`: How many distinct values are held in that object

-   `Mode`: What is the classification (type) of that object.

We will get to the "length" property later in the lesson.

The "`mode`" property corresponds to the type of data an object
represents.

The most common modes you will encounter in R are:

-   `numeric`: Numbers such floating point,decimals (1.0, 0.5, 3.14)

```{r numerics}
mode(1)
mode(1.1)
mode(1e6)
```

-   character A sequence of letters or numbers in single '' or double ""
    quotes

```{r character}
mode("a")
mode('a')
mode('1')
```

-   Logical Boolean values - `TRUE` or `FALSE`

```{r logicals}
mode(FALSE)
mode(TRUE)
```

There are a few other modes (i.e. "`complex`", "`raw`" etc.) but these
are the three we will work with in this lesson.

\
**Exercise: Create objects and check their modes**

```{r}
#Create the following objects in R, then use the mode() function to verify their modes. 
# Try to guess what the mode will be before you look at the solution

chromosome_name <- 'chr02'
od_600_value <- 0.47
chr_position <- '1001701'
spock <- TRUE
```

## Mathematical and functional operations on objects

R has a number of mathematical operators that you can use

-   addition `+`
-   subtraction `-`
-   multiplication `*`
-   division `/`
-   `^` or `**` exponentiation
-   a`%%`b modulus (returns the remainder after division)

These can be used with literal numbers:

```{r mathematical_operations}
(2^2)/1.5
```

And importantly, can be used on any object that evaluates to (i.e.
interpreted by R) a numeric object:

```{r}
# assign  numeric vale of 23 to object human_chr_number
human_chr_number =23
# multiply, * ,the object 'human_chr_number' by 2
human_chr_number * 2
```

## Vectors

**`Vectors`** are probably the most used commonly used object type in R.

A vector is a [collection]{.ul} of values that are [**all of the same
typ**]{.ul}**e** (numbers, characters, etc.).

One of the most common ways to create a vector is to use the

`c()` function - the "concatenate" or "combine" function.

Inside the function you may enter one or more values; for multiple
values, separate each value with a comma:

Create the vector `snp_genes` using these gene names "OXTR", "ACTN3",
"AR", "OPRM1" using the `c()` function

```{r create_vector}
# Create the SNP gene name vector with values "OXTR", "ACTN3", "AR", "OPRM1"
snp_genes<-c("OXTR", "ACTN3", "AR", "OPRM1")
```

Vectors always have a **mode** and a **length**. You can check these
with the

`mode()` and

`length()` functions respectively.

```{r mode_and_length}
# Check the mode, length, and structure of 'snp_genes'
mode(snp_genes)
length(snp_genes)
```

Another useful function that gives both of these pieces of information
is the `str()` (structure) function.

```{r str}
# Use the function str() on vector snp_genes
str(snp_genes)
```

## Creating and Subsetting vectors

Let's create a few more vectors to play around with:

-   snps that contains `rs53576, rs1815739, rs6152, rs1799971`

-   snp_chromosomes `3, 11, X, 6`

-   snp_positions `8762685, 66560624, 67545785, 154039662`

```{r create_vectors}
#Some interesting human SNPs
#while accuracy is important, typos in the data won't hurt you here

#snps object containing rs53576, rs1815739, rs6152, rs1799971
snps <- c('rs53576', 'rs1815739', 'rs6152', 'rs1799971')
# snp_chromosomes containing 3, 11, X, 6
snp_chromosomes <- c('3', '11', 'X', '6')
# snp _positions containing 8762685, 66560624, 67545785, 154039662
snp_positions <- c(8762685, 66560624, 67545785, 154039662)
```

## Retrieve values from a vector

Once we have vectors, one thing we may want to do is specifically
retrieve one or more values from our vector.

To do so, we use bracket notation.

We type the name of the vector followed by square brackets. `[]`

In those square brackets we place the index (e.g. a number) in that
bracket as follows:

[**In R index start from**]{.ul} 1

Get the 3rd value in the `snp_genes` vector

```{r subset_vector_by_index}
# get the 3rd value in the snp_genes vector
snp_genes[3]
```

#### Using Ranges 

In R, every item your vector is indexed, starting from the first item
(1) through to the final number of items in your vector.

You can also retrieve a range of numbers: `<start_num>:<end_num>`

```{r number_range}
# get the 1st through 3rd value in the snp_genes vector
snp_genes[1:3]
```

If you want to retrieve several (but not necessarily sequential) items
from a vector,

you pass a **vector of indices**; a vector that has the numbered
positions you wish to retrieve.

```{r}
# get the 1st, 3rd, and 4th value in the snp_genes vector

snp_genes[c(1, 3, 4)]
```

There are additional (and perhaps less commonly used) ways of subsetting
a vector (see [these
examples](https://thomasleeper.com/Rcourse/Tutorials/vectorindexing.html)).

Also, several of these subsetting expressions can be combined:

```{r}
# get the 1st through the 3rd value using range operator, and 4th value in the snp_genes vector
# yes, this is a little silly in a vector of only 4 values.
snp_genes[c(1:3,4)]
```

## Adding to, removing, or replacing values in existing vectors

Once you have an existing vector, you may want to add a new item to it.

To do so, you can use the `c()` function again to add your new value:

Add the gene `'CYP1A1'`and `'APOA5'`to our list of snp genes.

```{r add_data_to_vector}
# display current contents of snp_genes

# add the gene 'CYP1A1' and 'APOA5' to our vector of snp genes
# this overwrites our existing vector
snp_genes<-c(snp_genes,"CYP1A1","APOA5")

# use length() function on snp_genes()
length(snp_genes)
```

#### Negative indexes

Using a negative index `-` will return a version a vector with that
index's value removed:

Remove the sixth entry from snp_genes vector.

```{r }
#display contents of snp_genes and length()
# Remove the sixth entry from snp_genes vector, don't assign
snp_genes[-6]
# use length on snp_genes

```

We can remove that value from our vector by overwriting it with this
expression:

```{r}
# assign snp_genes to ovject snp_genes_bak
snp_genes_bak<-snp_genes
# remove 6th element of vector snp_gene sand store in snp_genes
snp_genes <- snp_genes[-6]
# display contents of snp_genes and use length() 
snp_genes
```

#### Assigning by index

We can also explicitly rename or add a value to our index using double
bracket notation `[]` and assigment operator `<-`

```{r vector_add_by_position}
# add gene name  "APOA5" ad the 7th item of the vector using [] notation
snp_genes[7]<- "APOA5"
# print contents and length
snp_genes
```

Notice in the operation above that R inserts an `NA` value to extend our
vector so that the gene "`APOA5`" is an index 7.

This may be a good or not-so-good thing depending on how you use this

## **Exercise: Examining and subsetting vectors**

Answer the following questions to test your knowledge of vectors

Which of the following are true of vectors in R?

A)  All vectors have a `mode` **or** a `length`\
B)  All vectors have a `mode` **and** a `length`\
C)  Vectors may have different lengths\
D)  Items within a vector may be of different modes\
E)  You can use the `c()` to add one or more items to an existing
    vector\
F)  You can use the `c()` to add a vector to an exiting vector

> \

## Logical Subsetting

`TRUE` `FALSE`

There is one last set of subsetting capabilities we want to introduce.

It is possible within R to retrieve items in a vector based on a logical
evaluation or numerical comparison.

E.g.

let's say we wanted get all of the SNPs in our vector of SNP positions
that were greater than 100,000,000.

We could index using the `>` (greater than) logical operator:

In the square brackets you place the name of the vector followed by the
comparison operator and (in this case) a numeric value.

`snp_positions[snp_positions > 100000000]`

Some of the most common logical operators you will use in R are:

    Operator    Description
    <           less than
    <=        less than or equal to
    >           greater than
    >=        greater than or equal to
    ==        exactly equal to
    !=        not equal to
    !x        not x
    a | b       a or b
    a & b       a and b

Create a Logical vector of snp_positions \> 100000000

```{r create_logical_vector}
# Create a Logical vector snp_pop_gt1M of snp_positions > 100000000
snp_positions
snp_positions > 100000000
```

Use function `which()` to get index of `snp_pop_gt1M` vector

```{r logical_index}
which(snp_positions > 100000000)
```

## Not Available' / Missing Values

Finally, there are a few other common retrieve or replace operations you
may want to know about.

First, you can check to see if any of the values of your vector are
missing (i.e. are `NA`).

The `is.NA()` function will return a logical vector with TRUE for any NA
value:

Find NA values value of `snp_genes` vector

```{r na_values}
# print current value of 'snp_genes': 
snp_genes

#use is.na function to find which values are NA
is.na(snp_genes)
```

## **Review Exercise 1**

What data types/modes are the following vectors?

a\. `snps`\
b. `snp_chromosomes`\
c. `snp_positions`

## **Review Exercise 2**

Add the following values to the specified vectors:

a\. To the `snps` vector add: 'rs662799'\
b. To the `snp_chromosomes` vector add: 11\
c. To the `snp_positions` vector add: 116792991

## **Review Exercise 3**

Make the following change to the `snp_genes` vector:

Hint: Your vector should look like this in 'Environment':
`chr [1:7] "OXTR" "ACTN3" "AR" "OPRM1" "CYP1A1" NA "APOA5"`. If not
recreate the vector by running this expression:
`snp_genes <- c("OXTR", "ACTN3", "AR", "OPRM1", "CYP1A1", NA, "APOA5")`

a\. Create a new version of `snp_genes` that does not contain `CYP1A1`
and then\

## **Extra List: Review Exercise 4**

Using indexing, create a new vector named `combined` that contains:

-   The the 1st value in `snp_genes`

-   The 1st value in `snps`

-   The 1st value in `snp_chromosomes`

-   The 1st value in `snp_positions`

## **Review Exercise 5**

What type of data is `combined`?

# Extra Material

## %in%

Sometimes, you may wish to find out if a specific value (or several
values) is present a vector.

You can do this using the comparison operator `%in%`

which will return `TRUE` for any value in your collection that is in the
vector you are searching:

Test to see if "ACTN3" or "APO5A" is in the snp_genes vector

if you are looking for more than one value, you must pass this as a
vector

```{r in_operator}
c("ACTN3","APOA5") %in% snp_genes
```

## Lists

Lists are quite useful in R, but we won't be using them in the genomics
lessons.

That said, you may come across lists in the way that some bioinformatics
programs may store and/or return data to you.

One of the key attributes of a list is that, unlike a vector, a list may
contain data of more than one mode.

In this one example, we will create a named list and show you how to
retrieve items from the list.

Create a named list using the `list()` function and our SNP examples

Note, for easy reading we have placed each item in the list on a
separate line

Nothing special about this, you can do this for any multiline commands

To run this command, make sure the entire command (all 4 lines) are
highlighted before running.

Note also, as we are doing all this inside the list() function use of
the '=' sign is good style

```{r create_list}
snp_genes<-c("OXTR", "ACTN3", "AR", "OPRM1")
snps <- c('rs53576', 'rs1815739', 'rs6152', 'rs1799971')
snp_chromosomes <- c('3', '11', 'X', '6')
snp_positions <- c(8762685, 66560624, 67545785, 154039662)

snp_data <- list(genes = snp_genes,
                 reference_snp = snps,
                 chromosome = snp_chromosomes,
                 position = snp_positions)
```

## List positions by name `$`

To get all the values for the position object in the list, we use the
`$` notation:

Return all the values of position object

```{r subset_list_by_name}
#Get position by name $
snp_data$position
mode(snp_data$position)
```

To get the first value in the position object, use the `[]` notation to
index:

```{r subset_list_by_name_and_index}
snp_data$position[1]
```

## Key Points

-   Effectively using R is a journey of months or years. Still you don't
    have to be an expert to use R and you can start using and analyzing
    your data with with about a day's worth of training

-   It is important to understand how data are organized by R in a given
    object type and how the mode of that type (e.g. `numeric`,
    `character`, logical, etc.) will determine how R will operate on
    that data.

-   Working with vectors effectively prepares you for understanding how
    data are organized in R.

## Exercises

<https://datacarpentry.org/genomics-r-intro/02-r-basics/index.html>
