---
title: 'R Genomics: Data Wrangling and Analyses with Tidyverser'
output:
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: inline
  markdown: 
    wrap: 72
---

[Aggregating and Analyzing Data with
dplyr](https://datacarpentry.org/genomics-r-intro/04-dplyr/index.html)

## Questions

-   How can I manipulate data frames without repeating myself?

## Objectives

-   Select certain columns in a data frame with the **`dplyr`** function
    `select`.

-   Extract certain rows in a data frame according to logical (boolean)
    conditions with the **`dplyr`** function `filter`.

-   Link the output of one **`dplyr`** function to the input of another
    function with the 'pipe' operator `%>%`.

-   Add new columns to a data frame that are functions of existing
    columns with `mutate`.

-   Use the split-apply-combine concept for data analysis.

-   Use `summarize`, `group_by`, and `count` to split a data frame into
    groups of observations, apply summary statistics for each group, and
    then combine the results.

-   Employ the 'split-apply-combine' concept to split the data into
    groups, apply analysis to each group, and combine the results.

-   Export a data frame to a .csv file.

# **Data manipulation using `dplyr`** and **`tidyr`** from **`tidyverse`** package

Bracket subsetting is handy, but it can be cumbersome and difficult to
read, especially for complicated operations.

Luckily, the [`dplyr`](https://cran.r-project.org/package=dplyr) package
provides a number of very useful functions for manipulating data frames
in a way that will reduce repetition, reduce the probability of making
errors, and probably even save you some typing.

As an added bonus, you might even find the `dplyr` grammar easier to
read.

Here we're going to cover some of the most commonly used functions as
well as using pipes (`|>`) to combine them:

1.  `glimpse()`

2.  `select()`

3.  `filter()`

4.  `group_by()`

5.  `summarize()`

6.  `mutate()`

Packages in R are sets of additional functions that let you do more
stuff in R.

The functions we've been using, like `str()`, come built into R;
packages give you access to more functions.

You need to install a package and then load it to be able to use it

```{r}
## installs dplyr package
## install readr package
install.packages("dplyr") 
install.packages("readr") 
```

## **What is dplyr?**

The package `dplyr` is a fairly new (2014) package that tries to provide
easy tools for the most common data manipulation tasks.

This package is also included in the [`tidyverse`
package](https://www.tidyverse.org/), which is a collection of eight
different packages

(`dplyr`, `ggplot2`, `tibble`, `tidyr`, `readr`, `purrr`, `stringr`, and
`forcats`).

It is built to work directly with data frames.

### **Loading .csv files in tidy style**

The Tidyverse's `readr` package provides its own unique way of loading
.csv files in to R using `read_csv()`, which is similar to `read.csv()`.

`read_csv()` allows users to load in their data

1.  faster,

2.  doesn't create row names, and

3.  allows you to access non-standard variable names (ie. variables that
    start with numbers of contain spaces),

4.  and outputs your data on the R console in a tidier way.

    In short, it's a much friendlier way of loading in potentially messy
    data.

```{r}
# Read in rawdata/combined_tidy_vcf.csv using the readr package's read_csv function
variants_tbl <-read_csv(file.path("rawdata","combined_tidy_vcf.csv"))

#display variants_tbl tibble 
variants_tbl

#data frame variants
variants
```

# Selecting columns and filtering rows

Using bracket notation, in order to pull out certain columns, we would
have to know which index referred to that column,

which took an extra step,

and if we had re-ordered our columns or added one in the middle, the
indexes in our script would suddenly refer to the 'wrong' column, which
might cause confusion!

As a refresher, here's how we pulled out columns using the bracket
notation:

```{r}
# get columns: sample_id, REF, ALT, DP
variants_tbl[, c(1, 5, 6, 12)]
```

But our code isn't very easy to understand and we'd need to see the
output to make sure we had the correct columns.

Enter `dplyr`!

#### **How to `select()`** columns

To select columns of a data frame, use dplyr's `select()` function.

The first argument to this function is the data frame (`variants`), and
the subsequent arguments are the columns to keep.

**Note, In tidyverse, e.g. dplyr we do not quote "" column names**

```{r dplyr_select}
#select sample_id, REF, ALT, DP from variants df
select(variants_tbl, sample_id, REF, ALT, DP)
```

This code is much easier to understand!

To select all columns except certain ones, put a "`-`" in front of the
variable to exclude it.

select all but CHROM column

```{r dplyr_select_negate}
#select all but CHROM column
select(variants_tbl, -CHROM)
```

## [selection helpers](https://noteable.edina.ac.uk/user/jggyj4w4chknqatfm2ddvp/rstudio/help/library/tidyselect/help/language)

`dplyr` also provides useful functions to select columns based on their
names.

-   [`starts_with()`](https://noteable.edina.ac.uk/user/jggyj4w4chknqatfm2ddvp/rstudio/help/library/tidyselect/help/starts_with):
    Starts with a prefix.

-   [`ends_with()`](https://noteable.edina.ac.uk/user/jggyj4w4chknqatfm2ddvp/rstudio/help/library/tidyselect/help/ends_with):
    Ends with a suffix.

-   [`contains()`](https://noteable.edina.ac.uk/user/jggyj4w4chknqatfm2ddvp/rstudio/help/library/tidyselect/help/contains):
    Contains a literal string.

-   [`matches()`](https://noteable.edina.ac.uk/user/jggyj4w4chknqatfm2ddvp/rstudio/help/library/tidyselect/help/matches):
    Matches a regular expression.

-   [`num_range()`](https://noteable.edina.ac.uk/user/jggyj4w4chknqatfm2ddvp/rstudio/help/library/tidyselect/help/num_range):
    Matches a numerical range like x01, x02, x03.

For instance, `ends_with()`

```{r dplyr_ends_with}
# select columns that end with B
select(variants_tbl, ends_with("B"))
```

## **Exercise**

From the `variants_tbl`

1.  Create a table that `contains` all the columns with the letter "`i`"
    and column "`POS`", [without]{.ul} columns "`Indiv`" and "`FILTER`".

Hint: look at the help function `contains()`

```{r}

```

## dplyr filtering

To choose rows, we can use dplyr's `filter()` function.

For instance, to keep the rows for the `sample_id` `SRR2584863`:

```{r dplyr_filter}
# use dplyr's filter funtion to keep the rows for the sample_id SRR2584863:
filter(variants_tbl,sample_id=="SRR2584863")

```

Note that this is equivalent to the base R code below, but is easier to
read!

```{r r_base_filter}
# base R code
variants_tbl[variants_tbl$sample_id == "SRR2584863",]
```

`filter()`will keep all the rows that match the conditions that are
provided.

Here are a few examples:

```{r dplyr_filter_examples}
## rows for which the reference genome has T or G 
filter(variants_tbl, REF %in% c("T", "G")) 

## rows with QUAL values greater than or equal to 100
filter(variants_tbl, QUAL >= 100)

## rows that have TRUE in the column INDEL
filter(variants_tbl,INDEL==TRUE)
```

`filter()` allows you to combine multiple conditions.

You can separate them using a comma `,` as arguments to the function,

they will be combined using the `&` (AND) logical operator.

If you need to use the `|` (OR) logical operator, you can specify it
explicitly:

```{r}
# filter variants_tbl
# filter variants_tbl using for sample_id SRR2584863 and QUAL score >= 100
sample_id == "SRR2584863" & QUAL >= 100)
filter(variants, sample_id == "SRR2584863", QUAL >= 100)
```

```{r}
# using `|` logical operator
# filter variants_tbl for sample_id SRR2584863 and (variants is INDEL or QUAL score >= 100)
filter(variants_tbl, sample_id == "SRR2584863", (INDEL | QUAL >= 100))
```

## **Exercise: dplyr filter**

`filter` for all the mutations in the `variants_tbl` that occurred
between

1.  the positions (`POS`) `1e6` (one million) and `2e6`(included)
2.  Alos have `QUAL` greater than 200.

```{r}

```

## Pipes

But what if you wanted to `select` and `filter`?

We can do this with pipes.

Pipes let you take the output of one function and send it directly to
the next

Pipes in R look like `|>` ; Previous only available via the `magrittr`
package, which is installed as part of dplyr using `%>%`.

Use pipes together with select and filter to get `REF`, `ALT`, `DP`
columns for `sample_id` `SRR2584863`

```{r pipe_example}
# Use pipes together with select and filter to get REF, ALT, DP columns 
# for sample_id SRR2584863
variants_tbl |>
  filter(sample_id == "SRR2584863")|>
  select(REF, ALT, DP)
```

We want to create a new object with this smaller version of the data we
can do so by assigning it a new name:

```{r pipe_example_new_variable}
# Use pipes together with select and filter to get REF, ALT, DP columns for sample_id SRR2584863 and save to new dataframe SRR2584863_variants
SRR2584863_variants <- variants_tbl |>
  filter(sample_id == "SRR2584863") |>
  select(REF, ALT, DP)

# print SRR2584863_variants
SRR2584863_variants
```

## **Exercise: Pipe and filter**

Starting with the `variants` dataframe, use pipes `|>` to subset the
data to include only observations from `SRR2584863` sample,

1.  Where the filtered depth (`DP`) is at least 10.
2.  Retain only the columns `REF`, `ALT`, and `POS`.

```{r exercise answer}

```

## Adding a column, Mutate

Frequently you'll want to create new columns based on the values in
existing columns,

for example to do unit conversions or find the ratio of values in two
columns.

For this we'll use the `dplyr` function `mutate()`.

We have a column titled "`QUAL`".

This is a Phred-scaled confidence score that a polymorphism exists at
this position given the sequencing data.

Lower QUAL scores indicate low probability of a polymorphism existing at
that site.

We can convert the confidence value QUAL to a probability value
according to the formula:

`Probability = 1- 10 ^ -(QUAL/10)`

Let's add a column (`POLPROB`) to our variants dataframe that shows the
probability of a polymorphism at that site given the data.

```{r adding_a_column_using_dplyr}
# add a column (POLPROB) to our variants dataframe using following formula: 
# 1 - (10 ^ -(QUAL/10)
variants_tbl %>%
  mutate(POLPROB = 1 - (10 ^ -(QUAL/10)))
```

## **Exercise**

There are a lot of columns in our dataset, so let's just look at the

`sample_id`, `POS`, `QUAL`, and `POLPROB` columns for now.

Add a line to the below code to only show those columns.

```{r exercise_answer}
variants %>%
  mutate(POLPROB = 1 - (10 ^ -(QUAL/10)))
```

### Extra

Let's create a new column, called "`indel_size`" that contains the size
difference between the our sequences and the reference genome.

The function, `nchar()` returns the number of letters in a string.

```{r}
variants_tbl %>%
  mutate(indel_size = nchar(ALT) - nchar(REF)) %>% 
  select(ALT,REF,indel_size)
```

```{r variants_indel}
variants_indel <- variants_tbl %>%
  mutate(
    indel_size = nchar(ALT) - nchar(REF),
    mutation_type = case_when(
      indel_size > 0 ~ "insertion",
      indel_size < 0 ~ "deletion",
      indel_size == 0 ~ "point"
    ))
```

## Split-apply-combine data analysis and the summarize() function

Many data analysis tasks can be approached using the
"`split-apply-combine`" paradigm:

1.  `split` the data into groups,
2.  `apply` some analysis to each group,
3.  and then `combine` the results. dplyr makes

This very easy through the use of the

`group_by()`function, which splits the data into groups.

When the data is grouped in this way the `summarize()` function can be
used to collapse each group into a single-row summary.

`summarize()`does this by applying an aggregating or summary function to
each group.

```{r}
# For example, if we wanted to group by sample_id and find the number of rows of data for each sample, we would do:
variants %>%
  group_by(sample_id) %>%
  summarize(n_observations = n())
```

Here the summary function used was `n()` to find the count for each
group, which we displayed in a column which we called `n_observations`.

We can also apply many other functions to individual columns to get
other summary statistics.

For example, we can use built-in functions like

`mean()`,

`median()`,

`min()`, and

`max()`.

These are called "built-in functions" because they come with R and don't
require that you install any additional packages.

```{r group_by}
# View the highest (max()) filtered depth (DP) for each sample (sample_id):
variants %>%
  group_by(sample_id) %>%
  summarize(max(DP))
```

View the lowest and highest filtered depth (DP) for each sample using
group_by and summarize:

```{r high_low_DP}
# View the lowest and highest filtered depth (DP) for each sample using group_by and summarize:
variants_indel %>%
  group_by(sample_id) %>%
  summarize(min(DP),max(DP))
```

## Exporting

We can export this new dataset using `readr` function `write_csv()`:

write updated object `variants` to a csv file in the data directory

```{r write_csv}
# write object variants_indel to file data/variants_indel.csv
  getwd()
  write.csv(variants_tbl, file.path("../","data","variants_processed.csv"))
```

### **Note: Handling Missing Values In Data**

By default, all **R functions operating on vectors that contains missing
data will return NA**.

It's a way to make sure that users know they have missing data, and make
a conscious decision on how to deal with it.

When dealing with simple statistics like the mean, the easiest way to
ignore `NA` (the missing data) is to use `na.rm = TRUE` (`rm` stands for
remove).

```{r}
mean(c(1,2,3,NA))
mean(c(1,2,3,NA),na.rm=T)
```

## Key Points

-   Use the `dplyr` package to manipulate dataframes.

-   Use `select()`to choose variables from a dataframe.

-   Use `filter()` to choose data based on values.

-   Use the pipe operator `|>` to join functions

-   Use `group_by()` and`summarize()` to work with subsets of data.

-   Use `mutate()` to create new variables.
