---
title: 'R Basics continued - factors and data frames'
output:
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: inline
  markdown: 
    wrap: 72
---

# R Basics continued - factors and data frames

-   How do I get started with tabular data (e.g. spreadsheets) in R?

-   What are some best practices for reading data into R?

-   How do I save tabular data generated in R?

## **Working with spreadsheets (tabular data)**

There are several ways to import data into R.

For our purpose here, we will focus on using the tools every R
installation comes with (so called "`base`" R)

to import a comma-delimited file containing the results of our variant
calling workflow.

We will need to load the sheet using a function called `read.csv()`.

```{r read_csv}
# Review the arguments of the read.csv() function
?read.csv()
```

Now let's read in the file `combined_tidy_vcf.csv`

```{r read_in_data}
# use getwd() to show where we are 

## read in a CSV file which we download at the start (rawdata/combined_tidy_vcf.csv) and assigning it to object 'variants'
variants <- read.csv(file.path("rawdata","combined_tidy_vcf.csv"))
```

Use `View()`function to bring up inbuilt data frame viewer

```{r view_data}
# View variants
View(variants)
```

**Note**: This data is variant call data for an sequencing experiments
examining mutation rates between hypermutable and non-hypermutable
strains of *E. coli* strains see more information
[here](https://datacarpentry.org/genomics-r-intro/02-data-prelude/index.html)

The data is now stored in a data frame object variants.

## Summarizing and determining the structure of a data frame.

1.  A data frame is the standard way in R to store tabular data.

2.  A data fame could also be thought of as a collection of vectors, all
    of which have the same length.

Using only two functions, we can learn a lot about out data frame
including some summary statistics as well as well as the "structure" of
the data frame.

Let's examine what each of these functions can tell us:

```{r summarise}
## Use summary() function on variants data frame
summary(variants)
```

Our data frame had 29 variables (columns), so we get 29 fields that
summarize the data.

The `QUAL`, `IMF`, and `VDB` variables (and several others) are
numerical data and so you get summary statistics on the

-   min

-   max

-   mean,

-   median

-   interquartile ranges (IQR).

Many of the other variables (e.g. `sample_id`) are treated as characters
data (more on this in a bit).

### Structure

Now, let's use the `str()` (structure) function to look a little more
closely at how data frames work:

```{r str}
# Use the str() function on the variants data frame.
str(variants)
```

Ok, thats a lot up unpack! Some things to notice.

-   the object type `data.frame` is displayed in the first row along
    with its dimensions, in this case 801 observations (rows) and 4
    variables (columns)

-   Each variable (column) has a name (e.g. `sample_id`).

    -   This is followed by the object mode (e.g. chr, int, etc.).

    -   Notice that before each variable name there is a `$`

We can use the `$` symbol to extract a single column of a data frame

`<object_name>$<column_name>`

To get the CHROM column

`variants$CHROM`

```{r}
# extract a REF column from the variants object using $ syntax
```

### Skimr

A really useful function is `skim()` from the
[skimr](https://cran.r-project.org/web/packages/skimr/index.html)
package. Packages are extra code libraries that extend the functionality
of R.

```{r}
# install package skimr
# load skimr library
# use skim fucntion of variants dataframe
install.packages("skimr")
library(skimr)
skim(variants)
```

## Introducing Factors

`Factors` are the final major data structure we will introduce in our R
genomics lessons.

`Factors` can be thought of as `vectors` which are specialized for
**categorical data**.

We can convert a character vector to factor using the function
`as.factor()`

```{r}
# convert REF column to factor
variants$REF<-as.factor(REF)

# use str function on column
str(variants$REF)
```

Extract the `factor` column REF to a new object using `$ syntax`

```{r}
# subset the fct column REF to a new object using $ and save it object REF
REF <- variants$REF
```

Let's look at the first few items in our factor using `head() function`:

```{r head}
# Let's look at the first few items in our factor using head() function:
head(REF)

```

# Plotting

A common uses for factors will be when you plot categorical values.

For example, suppose we want to know how many of our variants had each
possible reference nucleotide

(or nucleotide combination) in the reference genome? We could generate a
plot:

Plot `REF` using the base R's `plot()` function

```{r plotting}
# Plot REF using the plot() function
plot(REF)
```

### Extra Forcats

If you are using factors you should use the `forcats` package that has
really useful functions to manage factors. Extra (forcats
)[<https://forcats.tidyverse.org/>]

```{r forcats}
# load library forcats
# examples
plot(fct_lump(REF,n=5))
#
plot(forcats::fct_infreq(fct_lump(REF,n=5)))
```

Back to data frames

## Subsetting data frames

Next, we are going to talk about how you can get specific values from
data frames.

The subsetting notation is very similar to what we learned for vectors.

The key differences include:

-   Typically provide two values separated by commas:
    `data.frame[row, column]`

Subset the first row and column of variants data frame.

```{r}
# Subset the first row and 1st column of variants data frame using [<rows>,<columns>] notation.
variants[1,1]
```

In cases where you are taking a continuous range of numbers use a colon
between the numbers (`start:stop`, inclusive)

Subset the first row and column of variants data frame.

```{r}
# Subset the first row and column of variants data frame.
variants[1:5,1:5]
```

For a non continuous set of numbers, pass a vector using `c()`

```{r}
# Subset the  rows 1,4 and 9 and the first 5 column of variants data frame.
variants[c(1,4,9),1:5]
```

#### Index using the name of a column(s)

Index using the name of a column(s) by passing them as vectors using
`c() e.g`

```{r}
# extract the first 5 rows and column name POS
variants[1:5,c("POS")]
```

## Exercise Subsetting data frames and Factors

Try subsetting the data frame **variants**

1.  Retrieve the last row and last column
2.  Retrieve the `INDEL` column by name
3.  Retrieve the last item, 801th, in the INDEL column
4.  Convert the INDEL column to a factor and plot.

```{r}

```

## Subsetting using logical statements

We can use logical statements to filter out data frames as well

```{r}
# filter data for row with sample_id = SRR2584863
variants[variants$sample_id=="SRR2584863",]
```

Finally, in all of the subsetting exercises above, we printed values to
the screen.

You can create a new data frame object by assigning them to a new object
name

Create a new data frame from variants object containing only
observations from `sample_id` `SRR2584863`

```{r }
# Create a new data frame, SRR2584863_variants from variants object containing only observations from sample_id SRR2584863

SRR2584863_variants <-variants[variants$sample_id=="SRR2584863",]
```

Check the dimension of the data frame `dim()` `nrow()` and `ncol()`

```{r data_frame_dimensions}
# Check the dimension of the data frame dim() nrow() and ncol()
dim(SRR2584863_variants)
nrow(SRR2584863_variants)
ncol(SRR2584863_variants)
```

get a summary of the data frame `summary()`

```{r data_frame_summary}
# get a summary of the data frame summary()
summary(SRR2584863_variants)
```

## **Data frame bonus material: math, sorting, renaming**

Here are a few operations that don't need much explanation, but which
are good to know.

There are lots of arithmetic functions you may want to apply to your
data frame, covering those would be a course in itself (there is some
starting material
[here](https://swcarpentry.github.io/r-novice-inflammation/15-supp-loops-in-depth/)).

Our lessons will cover some additional summary statistical functions in
a subsequent lesson, but overall we will focus on data cleaning and
visualization.

You can use functions like

`mean()` `min()` `max()`

on an individual column.

Find maximum depth, DP, of all variants

```{r max_depth}
# Find maximum depth, DP, of all variants using max() function
max(variants$DP)
```

You can sort a data frame using the `order()` function:

Sort variants data frame by DP

```{r sort_data_frame}
# Sort variants data frame by DP
sorted_by_DP <- variants[order(variants$DP), ]
# view the first 6 rows using head
head(sorted_by_DP$DP)
```

You can rename columns: `colnames()`

Rename column sample_id to strain

```{r rename_columns}
# get names of columns using colnames() function
colnames(variants)

# Rename column sample_id to strain (use colnames(variants) == "sample_id") to filter column
colnames(variants)[colnames(variants) == "sample_id"] <- "strain"

#Check the new column names.
colnames(variants)
```

## Exercise, subsetting using logicals and maths opertations

1.  What is the maximum mapping quality `MQ` in the variants data frame.
2.  What is the mean depth `DP`in the variants data frame.
3.  Use logical filtering to extract those variants with the maximum
    `MQ`

```{r answers}

```

## Saving your data frame to a file

We can save data to a file. We will save our `SRR2584863_variants`
object to a .csv using the functions

-   `write.csv()`
-   `file.path()`

```{r write_csv}
#Save SRR2584863_variants object  to a file data/SRR2584863_variants.csv using write.csv
write.csv(x = SRR2584863_variants, file = file.path("../","data","SRR2584863_variants.csv"))
```

## Key Points

-   It is easy to import data into R from tabular formats However, you
    still need to check that R has imported and interpreted your data
    correctly

-   There are best practices for organizing your data (keeping it tidy)
    and R is great for this

-   Base R has many useful functions for manipulating your data, but all
    of R's capabilities are greatly enhanced by software packages
    developed by the community

## Exercises

Swirl exercises data frames and factors
